### PR TITLE
feat: Catch more unsafe numbers

### PR DESCRIPTION
**Pull request by hildjj**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Catch subnormals, numbers that are flushed to zero,
and overly-large integers in no-unsafe-values.

#### What changes did you make? (Give an overview)

Added three numeric cases to no-unsafe-values.js:
- Flushed to zero: the parsed value is zero, but the original text implies that zero was not intended.  Example: `1e-400`
- Overly-large integers: the original text looks like an integer (no decimal point or exponent), but the value is outside the range `[Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER]`.  `Example: 9007199254740992`
- Subnormals: these are numbers that have reduced precision in an IEEE-754 f64.  Example: `2.2250738585072009e-308`

#### Related Issues

Refs #68

#### Is there anything you'd like reviewers to focus on?

Please double-check my error messages for your preferences.

### CODE REVIEW

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:8]: Do we need to match the entire number sequence? If not, can you add some comments explaining what patterns these regexes are meant to match?
```suggestion
const SMALL_NUMBERS = [/^\.(\d+)$/u, /^((?:\d+\.)?\d+)e$/iu];
```

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:21]: I think we need to give folks a bit more here because I'm guessing a lot of people don't know what "subnormal" means.

```suggestion
			subnormal: "Unexpected subnormal number '{{value}}` found.",
```

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:43]: Just eliminating one-character variables.
```suggestion
						for (const pattern of SMALL_NUMBERS) {
							const match = txt.match(pattern);
							// If the digits are all 0, it's ok.
							if (match?.[1]?.match(/[1-9]/u)) {
```

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:69]: Expanding variable names.
```suggestion
						const buffer = new ArrayBuffer(8);
						const view = new DataView(buffer);
						view.setFloat64(0, node.value, false);
						const bigInt = view.getBigUint64(0, false);
						// Subnormals have an 11-bit exponent of 0 and a non-zero mantissa.
						if ((bigInt & 0x7ff0000000000000n) === 0n) {
```

**Review by nzakas — CHANGES_REQUESTED** [review]
Nice job on this. I just left a few notes mostly for cleanup.

File: src/rules/no-unsafe-values.js
Comment (hildjj) [reply src/rules/no-unsafe-values.js:8]: Those are intentionally not anchored.  I'll add some comments and examples inline there to make it more clear.

File: src/rules/no-unsafe-values.js
Comment (hildjj) [reply src/rules/no-unsafe-values.js:21]: Agree, almost nobody deals with these enough, which is why I think this particular lint is so valuable.  Perhaps a link to https://en.wikipedia.org/wiki/Subnormal_number as well?

File: src/rules/no-unsafe-values.js
Comment (hildjj) [reply src/rules/no-unsafe-values.js:8]: I've just rewritten this whole section to make it more clear and less error-prone, I hope.  I just translated the ABNF for `number` into an anchored regex.

File: src/rules/no-unsafe-values.js
Comment (nzakas) [reply src/rules/no-unsafe-values.js]: We generally don't put links in the error messages, but I think it would be good to add into the rule description. (At some point we'll create a page for each rule, and that would be a good place to go into more detail.)

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:24]: Can we include the number to make this a bit clearer?
```suggestion
				"The number '{{ value }}` will evaluate to zero.",
```

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:26]: We don't put URLs in messages as that makes the message more difficult to read.
```suggestion
				"Unexpected subnormal number '{{ value }}' found. Subnormal numbers are outside the safe range.",
```

File: tests/rules/no-unsafe-values.test.js
Comment (nzakas) [inline tests/rules/no-unsafe-values.test.js:1]: Can you add some tests for `unsafeZero`?

**Review by nzakas — CHANGES_REQUESTED** [review]

File: tests/rules/no-unsafe-values.test.js
Comment (hildjj) [reply tests/rules/no-unsafe-values.test.js:1]: Sorry, I had done that, but didn't get the file in the commit. :)

File: src/rules/no-unsafe-values.js
Comment (hildjj) [reply src/rules/no-unsafe-values.js]: Whoops, I didn't see this until I just sent the update.  Willfix.

File: src/rules/no-unsafe-values.js
Comment (hildjj) [reply src/rules/no-unsafe-values.js:25]: There's nothing interesting in the number.  It's either going to be exactly the same as the highlighted text (if we use txt) or 0 (if we use node.value)

File: tests/rules/no-unsafe-values.test.js
Comment (nzakas) [reply tests/rules/no-unsafe-values.test.js:1]: No worries, I do that all the time!

File: src/rules/no-unsafe-values.js
Comment (nzakas) [reply src/rules/no-unsafe-values.js:25]: That's true when people are using ESLint in an editor. People also use ESLint in their CI system, where it just outputs messages onto the terminal. In that case, having the value in the message is really helpful.

File: src/rules/no-unsafe-values.js
Comment (nzakas) [inline src/rules/no-unsafe-values.js:18]: Descriptions are meant to be one sentence. I think the original description was fine. The rest of these details should go in the README.

Comment (hildjj) [reply src/rules/no-unsafe-values.js:18]: Since this is going to modify the readme, I'm going to rebase on top of main.  Hopefully the patch isn't large enough that creates a problem.

**Comment by hildjj** [comment]
Sorry forgot to fix the value issue also.  On it.

**Comment by hildjj** [comment]
I made the rest of the errors consistent at the same time.  Tests modified.

**Review by nzakas — DISMISSED** [review]
LGTM. Waiting for a second review before merging.

File: src/rules/no-unsafe-values.js
Comment (mdjermanovic) [inline src/rules/no-unsafe-values.js:51]: Looks like some text is missing between "the" and "has"?

**Review by nzakas — APPROVED** [review]
Re-approving. Waiting for @mdjermanovic to verify his comments have been addressed.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
